File: src/main/java/com/turchinsky/entities/TrainHasScheduleAndRouteEntity.java
package com.turchinsky.entities;

import javax.persistence.*;
import java.sql.Timestamp;

@Entity
@Table(name = "train_has_schedule_and_route")
@IdClass(TrainHasScheduleAndRouteEntityPK.class)
public class TrainHasScheduleAndRouteEntity {

    @Id
    @Column(name = "train_id", nullable = false)
    int trainId;

    @Id
    @Column(name = "route_id", nullable = false)
    int routeId;

    @Id
    @Column(name = "departure_time", nullable = false)
    Timestamp departureTime;

    @Basic
    @Column(name = "ordered_seats")
    int orderedSeats;

    @Basic
    @Column(name = "free_seats")
    int freeSeats;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "train_id", referencedColumnName = "id", insertable = false, updatable = false)
    TrainEntity refTrainEntity;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "route_id", referencedColumnName = "id", insertable = false, updatable = false)
    RouteEntity refRouteEntity;

    public int getTrainId() {
        return trainId;
    }

    public void setTrainId(int trainId) {
        this.trainId = trainId;
    }

    public int getRouteId() {
        return routeId;
    }

    public void setRouteId(int routeId) {
        this.routeId = routeId;
    }

    public Timestamp getDepartureTime() {
        return departureTime;
    }

    public void setDepartureTime(Timestamp departureTime) {
        this.departureTime = departureTime;
    }

    public int getOrderedSeats() {
        return orderedSeats;
    }

    public void setOrderedSeats(int orderedSeats) {
        this.orderedSeats = orderedSeats;
    }

    public int getFreeSeats() {
        return freeSeats;
    }

    public void setFreeSeats(int freeSeats) {
        this.freeSeats = freeSeats;
    }

    public TrainEntity getRefTrainEntity() {
        return refTrainEntity;
    }

    public void setRefTrainEntity(TrainEntity refTrainEntity) {
        this.refTrainEntity = refTrainEntity;
    }

    public RouteEntity getRefRouteEntity() {
        return refRouteEntity;
    }

    public void setRefRouteEntity(RouteEntity refRouteEntity) {
        this.refRouteEntity = refRouteEntity;
    }

    @Override
    public String toString() {
        return "TrainHasScheduleAndRouteEntity{" +
                "trainId=" + trainId +
                ", routeId=" + routeId +
                ", departureTime=" + departureTime +
                ", orderedSeats=" + orderedSeats +
                ", freeSeats=" + freeSeats +
                '}';
    }

}

File: src/main/java/com/turchinsky/kt/TrainHasScheduleAndRouteEntity.kt
//package com.turchinsky.kt
//
//import javax.persistence.*
//
//@Entity
//@Table(name = "train_has_schedule_and_route", schema = "public", catalog = "sbb")
//@IdClass(TrainHasScheduleAndRouteEntityPK::class)
//open class TrainHasScheduleAndRouteEntity {
//    @get:Id
//    @get:Column(name = "train_id", nullable = false, insertable = false, updatable = false)
//    var trainId: Int? = null
//
//    @get:Id
//    @get:Column(name = "route_id", nullable = false, insertable = false, updatable = false)
//    var routeId: Int? = null
//
//    @get:Id
//    @get:Column(name = "departure_time", nullable = false)
//    var departureTime: java.sql.Timestamp? = null
//
//    @get:Basic
//    @get:Column(name = "ordered_seats", nullable = true)
//    var orderedSeats: Int? = null
//
//    @get:Basic
//    @get:Column(name = "free_seats", nullable = true)
//    var freeSeats: Int? = null
//
//    @get:ManyToOne(fetch = FetchType.LAZY)
//    @get:JoinColumn(name = "train_id", referencedColumnName = "id")
//    var refTrainEntity: TrainEntity? = null
//
//    @get:ManyToOne(fetch = FetchType.LAZY)
//    @get:JoinColumn(name = "route_id", referencedColumnName = "id")
//    var refRouteEntity: RouteEntity? = null
//
//    override fun toString(): String =
//            "Entity of type: ${javaClass.name} ( " +
//                    "trainId = $trainId " +
//                    "routeId = $routeId " +
//                    "departureTime = $departureTime " +
//                    "orderedSeats = $orderedSeats " +
//                    "freeSeats = $freeSeats " +
//                    ")"
//
//    // constant value returned to avoid entity inequality to itself before and after it's update/merge
//    override fun hashCode(): Int = 42
//
//    override fun equals(other: Any?): Boolean {
//        if (this === other) return true
//        if (javaClass != other?.javaClass) return false
//        other as TrainHasScheduleAndRouteEntity
//
//        if (trainId != other.trainId) return false
//        if (routeId != other.routeId) return false
//        if (departureTime != other.departureTime) return false
//        if (orderedSeats != other.orderedSeats) return false
//        if (freeSeats != other.freeSeats) return false
//
//        return true
//    }
//
//}
//
//class TrainHasScheduleAndRouteEntityPK : java.io.Serializable {
//    @get:Id
//    @get:Column(name = "train_id", nullable = false, insertable = false, updatable = false)
//    var trainId: Int? = null
//
//    @get:Id
//    @get:Column(name = "route_id", nullable = false, insertable = false, updatable = false)
//    var routeId: Int? = null
//
//    @get:Id
//    @get:Column(name = "departure_time", nullable = false)
//    var departureTime: java.sql.Timestamp? = null
//
//    override fun equals(other: Any?): Boolean {
//        if (this === other) return true
//        if (javaClass != other?.javaClass) return false
//        other as TrainHasScheduleAndRouteEntityPK
//
//        if (trainId != other.trainId) return false
//        if (routeId != other.routeId) return false
//        if (departureTime != other.departureTime) return false
//
//        return true
//    }
//
//    // constant value returned to avoid entity inequality to itself before and after it's update/merge
//    override fun hashCode(): Int = 42
//
//}
